arangod/Aql/Functions.cpp

```diff
@@ -1510,6 +1544,141 @@ AqlValue Functions::Like(arangodb::aql::Query* query,
 }
 
+/// @brief function SPLIT
+AqlValue Functions::Split(arangodb::aql::Query* query,
+                          transaction::Methods* trx,
+                          VPackFunctionParameters const& parameters) {
+  ValidateParameters(parameters, "SPLIT", 1, 3);
+
+  // cheapest parameter checks first:
+  int64_t limitNumber = -1;
+  if (parameters.size() == 3) {
+    AqlValue aqlLimit = ExtractFunctionParameterValue(parameters, 2);
+    if (aqlLimit.isNumber()) {
+      limitNumber = aqlLimit.toInt64(trx);
+    }
+    else {
+      RegisterInvalidArgumentWarning(query, "SPLIT");
+      return AqlValue(AqlValueHintNull());
+    }
+
+    // these are edge cases which are documented to have these return values:
+    if (limitNumber < 0) {
+      return AqlValue(AqlValueHintNull());
+    }
+    if (limitNumber == 0) {
+      VPackBuilder result;
+      result.openArray();
+      result.close();
+      return AqlValue(result);
+    }
+  }
+
+  transaction::StringBufferLeaser regexBuffer(trx);
+  AqlValue aqlSeparatorExpression;
+  if (parameters.size() >= 2) {
+    aqlSeparatorExpression = ExtractFunctionParameterValue(parameters, 1);
+    if (aqlSeparatorExpression.isObject()) {
+      RegisterInvalidArgumentWarning(query, "SPLIT");
+      return AqlValue(AqlValueHintNull());
+    }
+  }
+
+  AqlValueMaterializer materializer(trx);
+  AqlValue aqlValueToSplit = ExtractFunctionParameterValue(parameters, 0);
+
+  if (parameters.size() == 1) {
+    // pre-documented edge-case: if we only have the first parameter, return it.
+    VPackBuilder result;
+    result.openArray();
+    result.add(aqlValueToSplit.slice());
+    result.close();
+    return AqlValue(result);
+  }
+
+   // Get ready for ICU
+  transaction::StringBufferLeaser buffer(trx);
+  arangodb::basics::VPackStringBufferAdapter adapter(buffer->stringBuffer());
+  Stringify(trx, adapter, aqlValueToSplit.slice());
+  UnicodeString valueToSplit(buffer->c_str(), static_cast<int32_t>(buffer->length()));
+  bool isEmptyExpression = false;
+  // the matcher is owned by the query!
+  ::RegexMatcher* matcher = query->regexCache()->buildSplitMatcher(aqlSeparatorExpression, trx, isEmptyExpression);
+
+  if (matcher == nullptr) {
+    // compiling regular expression failed
+    RegisterWarning(query, "SPLIT", TRI_ERROR_QUERY_INVALID_REGEX);
+    return AqlValue(AqlValueHintNull());
+  }
+
+  VPackBuilder result;
+  result.openArray();
+  if (!isEmptyExpression && (buffer->length() == 0)) {
+    // Edge case: splitting an empty string by non-empty expression produces an empty string again.
+    result.add(VPackValue(""));
+    result.close();
+    return AqlValue(result);
+  }
+
+  std::string utf8;
+  static const uint16_t nrResults = 16;
+  UnicodeString uResults[nrResults];
+  int64_t totalCount = 0;
+  while (true) {
+    UErrorCode errorCode = U_ZERO_ERROR;
+    auto uCount = matcher->split(valueToSplit, uResults, nrResults, errorCode);
+    uint16_t copyThisTime = uCount;
```

Comment:
can't `uCount` be negative here? if so, what value will `copyThisTime` have then?